### DISCUSSION TITLE
product, permutations, combinations

### DISCUSSION THREAD

**Discussion by cjquines** [remeda/remeda#975] [Suggest a Function]
Related, but separate proposals:

- `product`, like [`xprod` from Ramda](https://ramdajs.com/docs/#xprod), [`itertools.product` from Python](https://docs.python.org/3.11/library/itertools.html#itertools.product), [`std::ranges::cartesian_product_view` from C++](https://en.cppreference.com/w/cpp/ranges/cartesian_product_view), [`itertools::Iterator::cartesian_product` from Rust](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.cartesian_product), requested in [Lodash](https://github.com/lodash/lodash/issues/852).
    - Would it be only for pairs, like `zip`, or variadic? It could be strictly typed too.
- `permutations`, like [`permutations` from Deno std](https://deno.land/std@0.212.0/collections/permutations.ts?s=permutations), [`itertools.permutations` from Python](https://docs.python.org/3.11/library/itertools.html#itertools.permutations), [`itertools::Iterator::permutations` from Rust](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.permutations), similar to [`std::next_permutation` from C++](https://en.cppreference.com/w/cpp/algorithm/next_permutation), requested in [Lodash](https://github.com/lodash/lodash/issues/1191).
    - Slightly different APIs here: do we only want permutations of the whole list, or permutations of a given size?
- `combinations`, like [`itertools.combinations` from Python](https://docs.python.org/3.11/library/itertools.html#itertools.combinations), [`itertools::Iterator::combinations` from Rust](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.combinations), requested in [Lodash](https://github.com/lodash/lodash/issues/1032).

**Comment by cjquines** [comment]
i also definitely have a lot more proposals like this that i can write up if that's something people want :shrug:

**Comment by eranhirsch** [comment]
These are all great and I don't mind adding them.

But be aware that their value in Python and other similar languages is that they rely heavily on the implementation using generators (yield) to make them efficient, as otherwise, they grow literally exponentially and could explode the memory in places where JS usually run which aren't optimized for memory-heavy computations like the browser and the edge.

Remeda has a notion of "iterator"-like API using our lazy evaluation mechanism, but it **only** works in pipes. This could trip users who use these methods naively in their dataFirst variant and we should avoid those situations:

```ts
for (const [a, b] of product(<BIG_ARRAY>, <BIG_ARRAY>)) {
  // The whole product array is first computed and loaded into memory
  // before an iterator over it is created by the engine...
  // Probably not what the user thought would happen
}
```

**Comment by eranhirsch** [comment]
@TkDodo - wdyt?
